### PR TITLE
sql/parser: Change Float -> Int cast from a trunctation to a round

### PR DESCRIPTION
**Pull request by nvb**
We currently handle integer casts from floats as truncations like they would be in Go. It looks like Postgres handles these in the same way (rounding to nearest even) as it would handle the ROUND builtin function with zero decimal places, before making the type change. See below. This is also how I'm currently planning on implementing the Decimal -> Int cast, unless a reason not to comes up.

```
nathan=# SELECT 1.5::FLOAT::INT;
 int4 
------
    2
(1 row)

nathan=# SELECT 2.5::FLOAT::INT;
 int4 
------
    2
(1 row)

nathan=# SELECT -1.5::FLOAT::INT;
 ?column? 
----------
       -2
(1 row)

nathan=# SELECT -2.5::FLOAT::INT;
 ?column? 
----------
       -2
(1 row)
```

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3901)

### CODE REVIEW

File: sql/parser/eval.go [hunk 1]
```diff
@@ -723,7 +723,11 @@ func (expr *CastExpr) Eval(ctx EvalContext) (Datum, error) {
 		case DInt:
 			return d, nil
 		case DFloat:
-			return DInt(v), nil
+			f, err := round(float64(v), 0)
+			if err != nil {
+				return DNull, err
```
Comment (tamird) [inline sql/parser/eval.go]: this should return `nil, err`; `DNull` is actually a valid `Datum`.

Comment (tamird) [inline sql/parser/eval.go]: can you write a test to exercise this case?

**Comment by tamird** [comment]
LGTM

**Comment by nvanbenschoten** [comment]
`round` should never return an error if given any float and `0` for its digits param, so it's probably best to panic instead of returning the error. Also panic-ing in `round` now because something is seriously wrong if `strconv.ParseFloat(strconv.FormatFloat(..))` gives an error.

File: sql/parser/builtins.go [hunk 2]
```diff
@@ -1260,7 +1260,10 @@ func round(x float64, n int64) (Datum, error) {
 	}
 	const b = 64
 	y, err := strconv.ParseFloat(strconv.FormatFloat(x, 'f', int(n), b), b)
-	return DFloat(y), err
+	if err != nil {
+		panic(fmt.Sprintf("parsing a float that was just formatted failed: %v", err))
```
Comment (tamird) [inline sql/parser/builtins.go]: by convention: %s for errors

File: sql/parser/eval.go [hunk 3]
```diff
@@ -723,7 +723,11 @@ func (expr *CastExpr) Eval(ctx EvalContext) (Datum, error) {
 		case DInt:
 			return d, nil
 		case DFloat:
-			return DInt(v), nil
+			f, err := round(float64(v), 0)
+			if err != nil {
+				panic(fmt.Sprintf("round should never fail with digits hardcoded to 0: ", err))
```
Comment (tamird) [inline sql/parser/eval.go]: missed the printf verb

**Comment by tamird** [comment]
LGTM

**Pull request merged**